File: practica 9/ObjetosColoreados.py
from abc import ABC, abstractmethod
import random
import math



class Coloreado:
    def como_colorear(self):

        pass

class Figura(ABC):
    def __init__(self, color=""):
        self.color = color

    def set_color(self, color):
        self.color = color

    def get_color(self):
        return self.color

    def __str__(self):
        return f"Figura de color {self.color}"

    @abstractmethod
    def area(self):

        pass

    @abstractmethod
    def perimetro(self):

        pass

class Cuadrado(Figura, Coloreado):
    def __init__(self, lado, color=""):
        super().__init__(color)
        self.lado = lado

    def area(self):
        return self.lado * self.lado

    def perimetro(self):
        return 4 * self.lado

    def como_colorear(self):
        return "Colorear los cuatro lados"

    def __str__(self):
        return f"Cuadrado con lado {self.lado} y color {self.color}"


class Circulo(Figura):
    def __init__(self, radio, color=""):
        super().__init__(color)
        self.radio = radio

    def area(self):
        return math.pi * self.radio ** 2

    def perimetro(self):
        return 2 * math.pi * self.radio

    def __str__(self):
        return f"Círculo con radio {self.radio} y color {self.color}"


def main():

    colores = ["rojo", "azul", "verde", "amarillo", "negro"]


    figuras = []


    for i in range(5):
        tipo_figura = random.randint(1, 2)
        color = random.choice(colores)

        if tipo_figura == 1:
            lado = random.uniform(1.0, 10.0)  # Lado aleatorio entre 1 y 10
            figura = Cuadrado(lado, color)
        else:
            radio = random.uniform(1.0, 10.0)  # Radio aleatorio entre 1 y 10
            figura = Circulo(radio, color)

        figuras.append(figura)


    print("INFORMACIÓN DE LAS FIGURAS ")
    for i, figura in enumerate(figuras, 1):
        print(f"Figura {i}: {figura}")
        print(f"Área: {figura.area():.2f}")
        print(f"Perímetro: {figura.perimetro():.2f}")


        if isinstance(figura, Coloreado):
            print(f"Método para colorear: {figura.como_colorear()}")
        else:
            print("Esta figura no implementa el método como_colorear")


if __name__ == "__main__":
    main()

File: practica 9/ObjetosColoreados.java
import java.util.Random;

interface Coloreado {
    String comoColorear();
}

abstract class Figura {
    protected String color;

    public Figura(String color) {
        this.color = color;
    }

    public abstract double area();
    public abstract double perimetro();
    public String getColor() {
        return color;
    }
    public void setColor(String color) {
        this.color = color;
    }
}

class Cuadrado extends Figura implements Coloreado {
    private double lado;

    public Cuadrado(double lado, String color) {
        super(color);
        this.lado = lado;
    }

    @Override
    public double area() {
        return lado * lado;
    }

    @Override
    public double perimetro() {
        return 4 * lado;
    }

    @Override
    public String comoColorear() {
        return "Colorear los cuatro lados";
    }

    @Override
    public String toString() {
        return "Cuadrado de lado " + lado + " y color " + color;
    }
}

class Circulo extends Figura {
    private double radio;

    public Circulo(double radio, String color) {
        super(color);
        this.radio = radio;
    }

    @Override
    public double area() {
        return Math.PI * radio * radio;
    }

    @Override
    public double perimetro() {
        return 2 * Math.PI * radio;
    }

    @Override
    public String toString() {
        return "Círculo de radio " + radio + " y color " + color;
    }
}

public class ObjetosColoreados {
    public static void main(String[] args) {
        String[] colores = {"rojo", "azul", "verde", "amarillo", "negro"};
        Figura[] figuras = new Figura[5];
        Random r = new Random();

        for (int i = 0; i < figuras.length; i++) {
            String color = colores[r.nextInt(colores.length)];
            if (r.nextBoolean()) {
                double lado = 1 + r.nextDouble() * 9;
                figuras[i] = new Cuadrado(lado, color);
            } else {
                double radio = 1 + r.nextDouble() * 9;
                figuras[i] = new Circulo(radio, color);
            }
        }

        for (int i = 0; i < figuras.length; i++) {
            System.out.println("Figura " + (i + 1) + ": " + figuras[i]);
            System.out.printf("Área: %.2f", figuras[i].area());
            System.out.printf("Perímetro: %.2f", figuras[i].perimetro());

            if (figuras[i] instanceof Coloreado) {
                System.out.println("Cómo colorear: " + ((Coloreado) figuras[i]).comoColorear());
            }
        }
    }
}
